bgpd/bgp_route.c

```diff
@@ -2821,6 +2821,23 @@ bool subgroup_announce_check(struct bgp_dest *dest, struct bgp_path_info *pi,
 					false));
 	}
 
+	/*
+	 * Adjust AIGP for propagation when the nexthop is set to ourselves,
+	 * e.g., using "set ip nexthop peer-address" or when advertising to
+	 * EBGP. Note in route reflection the nexthop is usually unmodified
+	 * and the AIGP should not be adjusted in that case.
+	 */
+	if (CHECK_FLAG(attr->flag, ATTR_FLAG_BIT(BGP_ATTR_AIGP)) &&
+	    AIGP_TRANSMIT_ALLOWED(peer)) {
+		if (nh_reset ||
+		    CHECK_FLAG(attr->rmap_change_flags,
```

Comment:
What about cases like `set ip next-hop unchanged`? Maybe we should just use a generic check `bgp_rmap_nhop_changed()`?

Reply:
These are the only two cases where the nexthop field is filled in by the local address, and the aigp should be adjusted.  No adjustment should be done if the nexthop remains unchanged.

If someone does "set ip next-hop x.x.x.x" (along with aigp) and uses one of the local addresses *other than* the peer-address, that can be TBD if needed.  But there needs to be some good reason for doing that.

Reply:
Correct, didn't check how nh_reset is derived.